### PR TITLE
Allow innerblocks to be null

### PR DESCRIPTION
**Pull request by scruffian**
There are some instances where the innerblocks of a block can be null. In these cases we need to check that they exist before we parse them etc.

The docblock of [this function](https://github.com/WordPress/WordPress/blob/c03305852e7e40e61cad5798eba9ebc3b961e27a/wp-includes/class-wp-block-list.php#L81) allows blocks to be null, which means that inner_blocks can also be null [here](https://github.com/WordPress/WordPress/blob/c03305852e7e40e61cad5798eba9ebc3b961e27a/wp-includes/class-wp-block.php#L153).

To reproduce this, add a social links block to your navigation block (using Gutenberg trunk). You should see a fatal error in the frontend of your site.

Trac ticket: 

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-block.php [hunk 1]
```diff
@@ -223,7 +223,7 @@ public function render( $options = array() ) {
 			foreach ( $this->inner_content as $chunk ) {
 				if ( is_string( $chunk ) ) {
 					$block_content .= $chunk;
-				} else {
+				} else if ( ! empty( $this->inner_blocks[ $index ] ) ) {
```
Comment (jmdodd) [inline src/wp-includes/class-wp-block.php]: I think a safer change is a check for specifically ! is_null.

For context:
`$this->inner_blocks` is assigned a new WP_Block_List; the offsetGet method (i.e. `$this->inner_blocks[ $index ]`) can return [either a block value, or null](https://github.com/WordPress/wordpress-develop/blob/0cb8475c0d07d23893b1d73d755eda5f12024585/src/wp-includes/class-wp-block-list.php#L89).

In PHP 8, trying to assign a property to null was escalated from a **warning** (Creating default object from empty value) to a **fatal** (Fatal error: Uncaught Error: Attempt to assign property "bar" on null).

I don't know under what circumstances an empty WP_Block_List object might be returned, but given how integral this function is to blocks working, I'd keep changes as small as possible.

**Comment by getdave** [comment]
Related https://github.com/WordPress/gutenberg/pull/46374

**Comment by dmsnell** [comment]
Hm. It seems like we have `inner_blocks` replaced with `Block_List`, which doesn't adhere to the specification?

those `$chunks` should only ever be string or `null` values. covering over the change here worries me that we're covering a domain model problem; if `inner_content` references more blocks than are in `inner_blocks` then we're going to have other problems.

sounds like we cut ourselves by adding an abstraction without implementing the specification: abstraction leak.

can we try to determine if the problem is in the implementation of `WP_Block_List` or in the social block? if we _ever_ modify `inner_blocks` we need to also modify `inner_content`

---

at a minimum, if we implement this workaround, can we at least add a `@TODO` indicating that it's covering up a deeper problem and we should come back and fix that problem, then remove the workaround, so we can avoid solving it multiple times in multiple ways? 😄

**Comment by scruffian** [comment]
@dmsnell I agree there is a separate issue in the social links block which is allowing `inner_blocks` and `inner_content` to not match. However it seems to me that that if we do ever get into that case it is better to fail gracefully than to fatal, which is what happens at the moment. What do you think?

**Comment by dmsnell** [comment]
> it is better to fail gracefully than to fatal, which is what happens at the moment. What do you think?

well to start with, this is why I asked, that if we apply this patch, that we add a comment indicating we've covered up the issue instead of fixing it, and also that we follow-up and fix the real problem and remove the workaround.

I don't always think it's better to fail gracefully because that brings up the question of what graceful means and to whom. presumably we found this issue because it fatal'd, and if we had covered it up earlier we wouldn't have noticed. this "fix" could also mean we're hiding bugs in other peoples' code as well.

it's not like I want to see sites crash, but I don't want us simply to see a flashing light, put tape over it, push data corruption further into the stack where it'll be harder to diagnose, and go on happy 😉

with the tag processor we've been using a somewhat hackish form of `assert()` which is to throw upon some condition if `WP_DEBUG` is set and silently fail if not. the goal is to quickly alert developers that what they are doing is wrong but then harm the end-user as little as we can, given the system failure.

So all this is to say: this patch as written will prevent some sites from crashing but might unintentionally invite worse problems that lead to more obvious data corruption issues on render, issues which will be hidden because we disabled the alarm.

I'm looking at WordPress/gutenberg#46279 and might add a comment there; I'm not familiar with peoples' use of directly modifying `inner_blocks` so I'm not sure what to say _in an urgent sense_ but I know we made a mistake by separating `inner_contents` and `inner_blocks` and the fix should probably involve re-joining those _or_ deferring the issue by _not modifying inner blocks_ (perhaps through a more normative render filter, which is how this general kind of operation was intended to work)

```php
function block_core_navigation_restrictor( $pre_render, $block ) {
	switch ( $block['blockName' ) {
		case 'core/navigation-link':
		case 'core/search':
		case 'core/social-links':
		case 'core/page-list':
		case 'core/spacer':
		case 'core/home-link':
		case 'core/site-title':
		case 'core/site-logo':
		case 'core/navigation-submenu':
			// don't touch these blocks
			return null;
	}

	// erase any other block
	return '';
}

function render_block_core_navigation( $attributes, $content, $block ) {
	// Prevent unintentional infinite loops and reject attempts to
	// use the nav block in a way we didn't prescribe or anticipate.
	add_filter( 'pre_render_block', 'block_core_navigation_restrictor', 10, 2 );

	$rendered = block_core_navigation_render_for_real( $attributes, $content, $block );

	remove_filter( 'pre_render_block', 'block_core_navigation_restrictor' );

	return $rendered;
}
```

A system like this doesn't mangle blocks and it lets us focus on the specific block type instead of requiring we also write our own recursive iterator to traverse the block tree.

**Pull request closed without merging**